File: QoSI/TP4/legion/Ruby/main.rb
require 'discordrb'
require 'yaml'
require 'net/http'
require 'uri'
require 'json'

# Load configuration from config.yaml
config = YAML.load_file('config.yaml')

bot = Discordrb::Bot.new token: config['credentials']['discord']

bot.ready do |event|
  puts 'Bot is ready! Setting up handlers...'
  puts "WebSocket status: #{bot.profile.status}"

  channel_id = config['channels']['zephyrus']
  channel = bot.channel(channel_id)
  unless channel
    puts 'Channel is not accessible with the provided ID.'
    bot.stop
  else
    startTime = Time.now.to_i

    # Send a ping message
    msg = channel.send_message('Ping?')
    sleep(1)  # Give some time for message delivery
    endTime = Time.now.to_i
    latency = (endTime - startTime) * 1000

    # Delete the ping message
    msg.delete

    puts "Measured Latency: #{latency}ms"

    uri = URI('https://api.ipify.org')
    response = Net::HTTP.get_response(uri)
    if response.is_a?(Net::HTTPSuccess)
      public_ip = response.body
      timestamp = Time.now.to_i * 1000
      json_message = {
        latency: format('%.2f', latency),
        public_ip: public_ip,
        hostname: 'zephyrus',
        lang: 'ruby',
        timestamp: timestamp
      }.to_json

      puts "Formatted JSON message: #{json_message}"
      # Send the detailed information message
      channel.send_message("```json\n#{json_message}\n```")
      puts "Message sent successfully to channel ID: #{channel.id}"

      # Stop the bot after sending the message
      bot.stop
    else
      puts "Error fetching public IP: #{response.code} #{response.message}"
      # Stop the bot if there is an error fetching the IP
      bot.stop
    end
  end
end

begin
  bot.run
rescue OpenSSL::SSL::SSLError => ssl_error
  puts "SSL Error occurred: #{ssl_error.message}"
  puts "Attempting to reconnect..."
  sleep(5)  # Wait 5 seconds before attempting to reconnect
  retry
rescue StandardError => e
  puts "An error occurred: #{e.message}"
ensure
  bot.stop
end

File: QoSI/TP4/legion/Lua/main.lua
local discordia = require('discordia')
local http = require('http')
local fs = require('fs')
local json = require("dkjson")

function httpGET(url, callback)
    coroutine.wrap(function()
        url = http.parseUrl(url)
        local req = (url.protocol == 'https' and https or http).get(url, function(res)
            local body = {}
            res:on('data', function(s)
                body[#body + 1] = s
            end)
            res:on('end', function()
                res.body = table.concat(body)
                coroutine.wrap(callback)(res)   
            end)
            res:on('error', function(err)
                coroutine.wrap(callback)(res, err)  
            end)
        end)
        req:on('error', function(err)
            coroutine.wrap(callback)(nil, err)   
        end)
    end)()
end

-- Load configuration from JSON file
local function loadConfig()
    local file = fs.readFileSync('config.json')
    local data, pos, err = json.decode(file, 1, nil)
    if err then
        error("Failed to load configuration: " .. err)
    end
    return data
end

local config = loadConfig()

local client = discordia.Client()

client:on('ready', function()
    print('Bot is ready! Setting up handlers...')
    print('WebSocket status:', client.user.status)

    local startTime = os.time()
    local channelId = config.channels.zephyrus  -- Accessing channel ID from the config
    local channel = client:getChannel(channelId)
    if not channel then error("Channel is not accessible with the provided ID.") end

    local msg = channel:send('Ping?')
    discordia.Clock():waitFor('', 3000)
    local endTime = os.time()
    local latency = (endTime - startTime) * 1000
    msg:delete()

    print('Measured Latency:', latency..'ms')

    httpGET('https://api.ipify.org', function(res, err)
        if err then
            print('Error:', err)
        else
            local publicIp = (res.body)
            local timestamp = os.time() * 1000
            local jsonMessage = json.encode({
                latency = string.format("%.2f", latency),
                public_ip = publicIp,
                hostname = 'zephyrus',
                lang = 'luvit',
                timestamp = timestamp
            })

            print('Formatted JSON message:', jsonMessage)
            channel:send('```json\n'..jsonMessage..'\n```')
            print('Message sent successfully to channel ID:', channel.id)
            client:stop()  -- Shutdown the bot after sending the message
        end
    end)
end)

local token = config.credentials.discord  -- Accessing token from the config
client:run('Bot ' .. token)
